src/highlighter.rs

```diff
@@ -0,0 +1,156 @@
+use crate::colorize::styled_output;
+use crate::colorize::STL;
+
+pub enum PotentialProblem {
+    Head,
+    Tail,
+    Unkn,
+    DistplayOnly,
+}
+
+enum ParseState {
+    BeforeHead,
+    Head,
+    AfterHead,
+
+    Tail,
+    TailLabel,
+    TailDirect,
+    TailIndirect,
+    TailPure,
+
+    AfterTail,
+    Comment,
+}
+
+pub fn code_highlighter(line: usize, source: &str, pp: PotentialProblem) {
+    let mut state = ParseState::BeforeHead;
+    let mut i = 0;
+    let source = source.replace('\t', "    ").replace('\r', "");
+    match pp {
+        PotentialProblem::Head | PotentialProblem::Tail | PotentialProblem::Unkn => {
+            styled_output(&line.to_string(), vec![STL::Red])
+        }
+        PotentialProblem::DistplayOnly => styled_output(&line.to_string(), vec![STL::BrightCyan]),
+    };
+    styled_output("\t", vec![STL::Normal]);
+    styled_output("| ", vec![STL::BrightCyan]);
+
+    while i < source.len() {
+        let c = source.chars().nth(i).unwrap();
+        (state, i) = match state {
+            ParseState::BeforeHead => match c {
+                ' ' => {
+                    styled_output(&c.to_string(), vec![STL::Normal]);
+                    (ParseState::BeforeHead, i + 1)
+                }
+                '#' => (ParseState::Comment, i),
+                _ => (ParseState::Head, i),
+            },
+            ParseState::Head => match c {
+                ' ' => (ParseState::AfterHead, i),
+                '#' => (ParseState::Comment, i),
+                _ => {
+                    let is_label = if let Some(label_identificator_index) = source.find(':') {
+                        let comment_index = if let Some(comment_index) = source.find('#') {
+                            comment_index
+                        } else {
+                            source.len()
+                        };
+                        label_identificator_index <= comment_index
+                    } else {
+                        false
+                    };
+                    match pp {
+                        PotentialProblem::Head => {
+                            styled_output(&c.to_string(), vec![STL::OnBrightRed])
```

Comment:
You are creating new allocation with `c.to_string()` while `styled_output` accepts `&str` as an argument